protocols/dcutr/src/behaviour_impl.rs

```diff
@@ -56,14 +57,17 @@ pub enum Event {
         remote_peer_id: PeerId,
         error: Error,
     },
+    DirectConnectionUpgradeTimedout {
+        remote_peer_id: PeerId,
+    },
 }
 
 #[derive(Debug, Error)]
 pub enum Error {
     #[error("Failed to dial peer.")]
     Dial,
     #[error("Failed to establish substream: {0}.")]
-    Handler(ConnectionHandlerUpgrErr<void::Void>),
+    Handler(UpgradeError<void::Void>),
```

Comment:
Previously, we emitted an error (and thus closed the connection) when we ran into a timeout. Why are we changing this behaviour now?

Reply:
Isn't the behaviour the same? In the sense that the connection was closed because of a timeout and we are bubbling it up? But after reading and replying on https://github.com/libp2p/rust-libp2p/pull/3307#discussion_r1064264754 I now see that Timeout makes sense being a variant on the `Error` itself of `DirectConnectionUpgradeFailed` rather than a variant of Event.

Reply:
> Isn't the behaviour the same? In the sense that the connection was closed because of a timeout and we are bubbling it up?

Unless I am mistaken, the timeout refers to a stream upgrade. The connection can/should still be intact at this stage. Something like `PendingUpgrade` should trigger this timeout on the remote but the connection underneath works perfectly fine.

It does raise the question whether we should close an entire connection at all just because one protocol failed to upgrade a stream. Just disabling that protocol (and handler) and letting the collaborative keep-alive eventually close the connection seems like a better idea at first sight.

We may want to do this in a separate PR though, either before or after we merge this.